### PR TITLE
File refactoring

### PR DESCRIPTION
**Pull request by susgo**
I agree with @Fonata that the class file is very long. So in addition to #114 my proposal with this pull request is to exclude also some of the existing features into traits. Understanding the code seems a little bit easier for me after that. I know there should be also some properties excluded (not done at the moment)

Please feel free to close this pull request if this fits not your future ideas for this package.

Some other things I would like to change if you are interested in:
1. change visibility level of `parse_string` to `protected` for secruity reasons in user handling only (I have not changed it because test fails after changing)
2. add `Readme` for `tests`
3. refactor some functions as they are partially very long too
3. introduce some constants for some bools and strings which are used, e.g.:
    * for `append` in save method
    * for `sort_type`
    * for `encoding`
    * for `operators` in method `_validate_row_condition`

What do you think?

Thanks. :blush:

### CODE REVIEW

**Comment by Fonata** [comment]
I like the direction this is heading. Good job 👍 
Please add the suffix "Trait" to both the filenames and the trait names.

**Comment by Fonata** [comment]
Regarding trait files, `require` and `Bootstrap.php`: I firmly believe that a lot of existing users of this project will continue to use `require "parsecsv.lib.php";`. We must make sure that to them, a single ``require`` will continue to be enough.

Additionally, to keep this project future-proof, the Composer-configured autoloader has to just work, too.

**Comment by susgo** [comment]
Regarding your last point: What do you think should have to bee done?

**Comment by susgo** [comment]
Oh and what's your opinion on the other future improvement points?

**Comment by Fonata** [comment]
@susgo I'll go through your suggestions from above:

1. I'm a bit slow: can you explain why ``parse_string`` should be protected for "security reasons  in user handling"? 
2. absolutely. Running and extending tests can certainly be explained in a Readme file. Markdown?
3. I agree, refactoring is welcome. This pull request tends to become long already. I hate to slow you here; please consider doing that this pull request is completed so changes are easier to follow. If you are unsure if you break things: let's write some more tests first. Feel free to create specific issues if you want to give me a chance to take a share of the workload.
4. Backwards compatibility is my main priority with this project. I thought about NICE_CONSTANTS too, but the values need to remain identical so that people using this project don't have to update their code. 

Regarding Composer: once we move files to a sub-directory we can do either (but not both) of the following:
- Add a namespace segment with the same name as the sub-directory, or
- (less preferred) add the new dir to `classmap`.

The tricky part with the namespace is to come up with something that is not repetitive and favored by the community. Feel free to suggest or endorse something in #111.

As part of the 1.0.0 version, I now realized that for poeple using `require "parsecsv.lib.php";` the `use` statement should be in that traditional .lib.php file, but the class should be in a separate file that has no ".lib." in the file name.

Do ask if I mixed too many things together 😊

**Comment by susgo** [comment]
@Fonata: 
1. Maybe I miss something but I don't think a situation where people benefit from calling it directly instead of calling `parse`. Also we can nearly exclude possible faults in calling, so e.g. some comments in #89 are unnecessary than. All in all: I see no reason why it has to be public. But I'm not so familiar with all use cases, so maybe it's okay.

2. Of course Markdown.

3. I agree. My proposal is that you write down which changes you do need to merge this request and I go on with my refactoring work when merge is done.

4. That is also always my purpose.

Regarding composer: I have added the new directory to classmap but we should use namespaces instead - in my opinion. Decision should be done by a more experienced software developer than I am.

**Comment by Fonata** [comment]
Everything :+1: + some additional notes:
- I totally agree that `parse` should be people's choice. `parse_string` is less buggy than it used to and the confusion around having two functions alone is enough of a reason to make the less useful one (`parse_string`) as protected; however we know from the issues that some people already use it, so it would break things.  
TL;DR: we push it into version 1.0.0 as the semantic versioning should tell people to read the release notes for breaking changes. A few might actually do that 😛 
- Composer: Yes, we take the namespace approach. My main reason: the traits sub-dir. Plus: Namespaces are a PSR requirement.

**Comment by susgo** [comment]
I would suggest that we only include the namespace change into this branch. It's more easy to do all refactoring work in another branch - for the reviewer and also for me. Hope you agree.

**Comment by Fonata** [comment]
Hi **susgo**, I do agree. I'll continue the namespace discussion in #111

**Comment by Fonata** [comment]
Since you mentioned enums in the other pull request: To me it is of uttermost importance that we keep compatibility. For everything that existed before Dec 2017, the values of the constants have to be the exact same strings.

**Comment by susgo** [comment]
I know and I don't want to change the strings. I always use enums to group constants together as it is the cleanest way for introducing constants in a software - in my opinion only.

**Comment by Fonata** [comment]
I agree 100 % percent.

**Comment by susgo** [comment]
I close this one for the moment and rework it with the comments in #111 in another pull request.
